### PR TITLE
initial support for credential providers.

### LINKED ISSUE #1509: support for credential providers
Credential providers a la [Nuget 3](http://docs.nuget.org/Consume/Credential-Providers) allow for seemless authentication with private package feeds while improving security (no need to store password/token on disk). It would be great to add support for this in Paket.

### LINKED ISSUE #3021: Use of private VSTS Feed with a Azure AD
### Description

We have visual studio solutions that we want to migrate to VSTS.
These solutions use paket instead of nuget, and it work well today in our environment.

When we use paket inside a VSTS Build, paket do restore only packages from nuget.org.
But, these that come from a private VSTS nuget Feed are not resolved.

If I call paket restore from my dev machine, I have the same problem : only nuget are restored.
The attached file show the log in verbose mode for paket restore
[paket-error.txt](https://github.com/fsprojects/Paket/files/1682504/paket-error.txt)
(I have replace the real vsts name by xxxxx)

For information, this is a corporate VSTS, that need a CredentialProvider.VSS.exe when a nuget.exe (that works well).

### Repro steps

paket restore on a VSTS private feed, which need a nuget credential provider (if used)

### Known workarounds

no workarround for the moment

### PR DESCRIPTION
**Pull request by matthid**
closes #3021, closes #1509 

paket-credential-providers

# Paket support for Credential Providers

Paket supports Credential Providers through the same interface as [NuGet](https://docs.microsoft.com/en-us/nuget/reference/extensibility/nuget-exe-credential-providers).

Paket extends the definition and allows to run FDD-NetCore applications to run as well (`*.dll` files!), so Paket additionally searches for `CredentialProvider*.dll` files in the given paths. For this to work Paket needs to be able to resolve a `dotnet` executable from the `PATH` variable.

## Development

For regular paket users installing Credential Providers works the same as for the NuGet client. If you have already installed Credential Providers in `%LOCALAPPDATA%\NuGet\CredentialProviders` paket should pick them up immediatly.

Example VSTS:

1. Download the credential providers from your VSTS Instance.
   credential-providers-vsts

2. Extract the `CredentialProvider.VSS.exe` file into the above path (`%LOCALAPPDATA%\NuGet\CredentialProviders`)
3. use paket normally (without password in `paket.dependencies` and config) and enter the password in the provided dialog.

## CI

There are two options to use Credential Providers in your build agent:

1. Either install a global Credential Provider on your agent
2. Use Tasks to provide Credential Providers as part of your build.

Example VSTS:

Install https://github.com/matthid/Paket.TeamBuildCredentials/releases/tag/0.1.1 on your TFS
Add the "Setup Paket credential manager" build step before calling Paket.
See more infos [here](https://github.com/matthid/Paket.TeamBuildCredentials)

> Note: Disable failing on standard error for your build step calling paket.
> paket-vsts-disable-stderr

### CODE REVIEW

File: src/Paket.Core/PackageManagement/NugetConvert.fs [hunk 1]
```diff
@@ -28,19 +28,19 @@ type CredsMigrationMode =
 
     static member ToAuthentication mode sourceName auth =
         match mode, auth with
-        | Encrypt, Credentials(username, password, authType) -> 
-            ConfigAuthentication(username, password, authType)
-        | Plaintext, Credentials(username, password, authType) -> 
-            PlainTextAuthentication(username, password, authType)
```
Comment (matthid) [inline src/Paket.Core/PackageManagement/NugetConvert.fs]: This change is a bit fishy. I have no idea where this information was used (or how it was supposed to be used). Apparently this information was not used at all or I introduced a bug.

File: src/Paket.Core/PackageManagement/NugetConvert.fs [hunk 2]
```diff
@@ -28,19 +28,19 @@ type CredsMigrationMode =
...
             let question =
                 sprintf "Credentials for source '%s': " sourceName  + 
                     "[encrypt and save in config (Yes) " + 
                     sprintf "| save as plaintext in %s (No)]" Constants.DependenciesFileName
                     
             match Utils.askYesNo question with
-            | true -> ConfigAuthentication(username, password, authType)
-            | false -> PlainTextAuthentication(username, password, authType)
+            | true -> Credentials userPass
+            | false -> Credentials userPass
```
Comment (matthid) [inline src/Paket.Core/PackageManagement/NugetConvert.fs]: What was the effect of this?

**Comment by matthid** [comment]
This PR pretty much makes it possible to use existing NuGet credential providers (https://docs.microsoft.com/en-us/nuget/reference/extensibility/nuget-exe-credential-providers).

This works already quite nicely on the development PC, but I have no idea how we get a working story for VSTS Team Build. Any idea @isaacabraham ?

**Comment by isaacabraham** [comment]
@matthid working story? It should just work now I guess. Perhaps @foxrough can test it out to confirm though?

**Comment by matthid** [comment]
@isaacabraham I tested it and it doesn't just work. The nuget vsts task is doing a lot of magic internally and brings it's own CredentialProvider to make it work.

**Comment by matthid** [comment]
Problem is that doesn't quite fit in our model of bootstrapping paket within the build as well as how FAKE5 is working.

**Comment by matthid** [comment]
My current thinking is:
 - provide our own VSTS Extension with two tasks
 - One task sets up the credential provider (and needs to be before running the buildscript/paket) and saves the access token somewhere
 - One task cleans the credential provider

This would work for Paket and FAKE5

**Comment by isaacabraham** [comment]
@matthid I already did a (very simple) Paket and Fake extension in VSTS through the F# Helpers extension (https://marketplace.visualstudio.com/items?itemName=isaacabraham.fsharp-helpers-extension). However it's very simple and probably needs a rewrite if it's going to do anything serious. I'm happy to make you an admin of that repo if you want.

Alternatively, perhaps we can look / speak to the NuGet team to see how they're doing it in VSTS?

**Comment by matthid** [comment]
@isaacabraham Yes I can try to extend there, however the question is more fundamental. I'm pretty sure I know what they are doing, however our workflow looks a bit different.

But yes maybe extending your tasks to provide an credential manager (just like nuget does it) makes a lot more use-cases viable. Currently I'm trying around various possibilities and creating new tasks. But I can add them to your suite after they start working.

**Comment by matthid** [comment]
@isaacabraham I guess that is not best practice but I added https://github.com/matthid/Paket.TeamBuildCredentials and some documentation around it. 
It is not best practice because we store the OAuth token temporarily such that it is available for Paket. But I like that workflow a bit better.

We can add the tasks to your extension if you are OK with how they work. I tested a bit and everything seems to work just fine.

File: build.fsx [hunk 3]
```diff
@@ -75,6 +75,9 @@ let buildMergedDir = buildDir @@ "merged"
 let paketFile = buildMergedDir @@ "paket.exe"
 
 Environment.CurrentDirectory <- __SOURCE_DIRECTORY__
+
+System.Net.ServicePointManager.SecurityProtocol <- unbox 192 ||| unbox 768 ||| unbox 3072 ||| unbox 48
```
Comment (haf) [inline build.fsx]: You should avoid magic number and assign them to constants before ORing them.

Comment (matthid) [reply build.fsx]: @haf Yes good eyes ;) 
I quickly copy&pasted it from `Utils.fs`. Basically, I think we need to fix this in either Octokit or FAKE

**Pull request merged**